### DISCUSSION TITLE
Converting a non-ADmaterial to an ADmaterial

### DISCUSSION THREAD

**Discussion by mahyarmalek** [idaholab/moose#20926] [Q&A Modules: Solid mechanics] [answered]
Hello Dear Moose Developers 

I recently have inroduced to this amazing framework nd I'm researching on soil mechanics. I've noticed that there are a lot of plasticity models available in Moose but so many of them do not support the AD properties. Is there anyway to convert those plasticity models into AD? if yes, where can I start? what do I have to change? I wanted to use the crystal plasticity model but since my kernels are all AD and my displacements are coupled with phasefield and pressure and temperature it is impossible for me to use the non AD kernels or materials 

Best,

Mahyar

**Comment by GiudGiud** [comment, marked as answer]
Hello

You can combine AD kernels with non AD kernels, that is not an issue. AD materials and non-AD materials are harder to mix, but you can use this object
https://mooseframework.inl.gov/source/materials/MaterialADConverter.html
to create the non AD versions. 
One thing to be careful about though, is that the derivatives may not have been defined in the AD version of a material since they are automatically calculated. In this case, you will have to redefine them yourself.

Guillaume

**Reply by mahyarmalek** [reply]
Thank you for your answer, I'm trying to implement stress divergence which is multiplied by phasefield variable. My phasefield variable is AD variable, but I want to use crystal plasticity which is a non-AD material. So the kernel will be multiplication of a AD variable to a non-AD variable is that ok? Also, there are no examples on how to use the ADMaterialConverter. I completely new to this framework and really dont know how to use this function to convert material type. 

Best,

**Reply by GiudGiud** [reply]
There's not really such a thing as an AD variable. AD quantities are quantities that carry their derivative with regards to the nonlinear variables along with them.

Multiplying AD and non AD quantities is OK as long as:
- numerically, it turns out not to matter to be missing some Jacobian terms.
OR
- you make sure the contribution to the Jacobian is correct by implementing the computeQpJacobian routine yourself. This is disabled for ADKernels though. Implementing this could be just as much work as implementing an AD plasticity model.
What kernel are you using? Which plasticity model are you using?

Here's an example of the materialADConverter.
https://github.com/idaholab/moose/blob/8b86c661f83d7d0d6bfd97f78e949e6228c86d3c/test/tests/materials/ad_material/conversion/1d_dirichlet.i

**Reply by mahyarmalek** [reply]
Thank you for your answer, my kernel is an AD kernel written by myself for calculating the multi-phase linear momentum equation. It has three different stress divergence in it with three different base name. For the stress with the base name "ice" im trying to use the crystal plasticity model which is not an AD model. 

The problem is my linear momentum equation is coupled with phase-filed variable which is also an AD quantity. As far as I understood, when I have a AD kernel with all AD material properties, it is impossbile to use a non-AD material in between with a Newton solver that im using. 

fortunately you mentioned that I can still use a non-AD material and get it to work. I tried to change the material property to a non AD for the ice stress but for some reason crystal plasticity is not accepting the deformation gradients from the ComputeFiniteStrain (I asked this question in another thread). I dont think I can use the tensor mechaincs master class action since it just adds an stress divergence but my stress is multiplied by a phasefield variable which makes it impossible to use master class. 

Also I've tried to use the materialADConverter, but it gives me error that this object is not a registered object?? not sure what is this....

**Reply by GiudGiud** [reply]
> Thank you for your answer, my kernel is an AD kernel written by myself for calculating the multi-phase linear momentum equation. It has three different stress divergence in it with three different base name. For the stress with the base name "ice" im trying to use the crystal plasticity model which is not an AD model.

Ok an ADKernel will not let you write in the derivatives manually (well it could but we dont recommend doing it the way it could).
So you can either:
- work with converted AD properties (converted to AD using the converter, but missing the derivatives) and miss the derivative terms
- convert the crystal plasticity model. Which one is that? What is the object name?

> The problem is my linear momentum equation is coupled with phase-filed variable which is also an AD quantity. As far as I understood, when I have a AD kernel with all AD material properties, it is impossbile to use a non-AD material in between with a Newton solver that im using.

It is not impossible. It is just that using regular properties will miss some derivatives, so the Jacobian in the Newton method will not be numerically perfect. 

> fortunately you mentioned that I can still use a non-AD material and get it to work. I tried to change the material property to a non AD for the ice stress but for some reason crystal plasticity is not accepting the deformation gradients from the ComputeFiniteStrain (I asked this question in another thread). I dont think I can use the tensor mechaincs master class action since it just adds an stress divergence but my stress is multiplied by a phasefield variable which makes it impossible to use master class.

Ok makes sense why you cannot use the action.

> Also I've tried to use the materialADConverter, but it gives me error that this object is not a registered object?? not sure what is this....

can you paste it here? It must be mispelled. This is available in every moose that is less than 6 months old

**Reply by mahyarmalek** [reply]
> can you paste it here? It must be mispelled. This is available in every moose that is less than 6 months old

 I noticed my version is old, I added the specific code and compiled and it works. 
____________________________
> work with converted AD properties (converted to AD using the converter, but missing the derivatives) and miss the derivative terms
convert the crystal plasticity model. Which one is that? What is the object name?

I have tried it with converter and it's not converging, I think it's too nonlinear and if not provided with derivative it cannot converge. 
I am using the CrystalPlasticityKalidindiUpdate within ComputeMultipleCrystalPlasticityStress. I would assume the last solution is to write an AD material property for it but I really have no idea how to start the whole code is very overwhelming and hard to digest for me since i'm both new to framework and C++ . If you had any guidlines on how I can write an AD material code for plasticity and updating the stress it would be really helpful. 

Best, 

Mahyar

**Reply by GiudGiud** [reply]
so to converge a kernel or a material property to AD, in a quick way, not for general use in the framework, just copy paste the file with a different name and replace all instances of Real with ADReal. 

If you want to contribute to the framework, you ll have to make a template on the AD type. We can help you with that too
